tests/dataset_loader.py

```diff
@@ -0,0 +1,177 @@
+        self._default_df_type = df_type
+        self.__instances[df_type] = self
+
+    def get_alt_df_type(self, df_type: typing.Literal['cudf', 'pandas']) -> typing.Literal['cudf', 'pandas']:
+        """Returns the other possible df type."""
+        return 'cudf' if df_type == 'pandas' else 'pandas'
+
+    def clear(self):
+        self.__df_cache.clear()
+
+    def get_df(self,
+               file_path: str,
+               df_type: typing.Literal['cudf', 'pandas'] = None) -> typing.Union[cdf.DataFrame, pd.DataFrame]:
+        """
+        Fetch a DataFrame specified from `file_path`. If `file_path` is not an absolute path, it is assumed to be
+        relative to the `test/tests_data` dir. If a DataFrame matching both `file_path` and `df_type` has already been
+        fetched, then a cached copy will be returned. In the event that a DataFrame matching `file_path` but not
+        `df_type` exists in the cache, then the cached copy will be cast to the appropriate type, stored in the cache
+        and then returned
+        """
+        if os.path.abspath(file_path) != os.path.normpath(file_path):
+            full_path = os.path.join(TEST_DIRS.tests_data_dir, file_path)
+        else:
+            full_path = file_path
+
+        if df_type is None:
+            df_type = self._default_df_type
+
+        df = self.__df_cache.get((df_type, full_path))
+        if df is None:
+            # If it isn't in the cache, but we have a cached copy in another DF format use it instead of going to disk
+            alt_df_type = self.get_alt_df_type(df_type=df_type)
+            alt_df = self.__df_cache.get((alt_df_type, full_path))
+            if alt_df is not None:
+                if alt_df_type == 'cudf':
+                    df = alt_df.to_pandas()
+                else:
+                    df = cdf.DataFrame.from_pandas(alt_df)
+            else:
+                df = read_file_to_df(full_path, df_type=df_type)
+
+            self.__df_cache[(df_type, full_path)] = df
+
+        return df.copy(deep=True)
+
+    def __getitem__(
+        self, item: typing.Union[str, typing.Tuple[str], typing.Tuple[str, typing.Literal['cudf', 'pandas']]]
+    ) -> typing.Union[cdf.DataFrame, pd.DataFrame]:
+        if not isinstance(item, tuple):
+            item = (item, )
+
+        return self.get_df(*item)
+
+    @classmethod
+    def get_loader(cls, df_type: typing.Literal['cudf', 'pandas']):
+        """
+        Factory method to return an instance of `DatasetLoader` for the given df_type, returns `self` if the df_type
+        matches. Used by cudf and pandas propery methods.
+        """
+        try:
+            loader = cls.__instances[df_type]
+        except KeyError:
+            loader = cls(df_type=df_type)
+
+        return loader
+
+    @property
+    def cudf(self):
+        return self.get_loader(df_type='cudf')
+
+    @property
+    def pandas(self):
+        return self.get_loader(df_type='pandas')
```

Comment:
In the usage examples in the issue `DatasetLoader.cudf` was being used as a property:
```py
# Get specific pandas or cudf version
df_cudf = dataset.cudf["filter_probs.csv"]
df_pandas = dataset.pandas["filter_probs.csv"]
```

The `classmethod` and `property` decorators can't be combined in python 3.8. This was actually added in 3.9 but then removed in 3.11. 

It looks like it could be done with a little effort by defining them in a metaclass, which would add the properties to class but not to instances.
We could then add them to the instances in the `__new__` method, but I'm just not sure how much benefit there is to having these as class properties.